File: impl/ts/src/tool/validate.ts
// Shared helpers for tool arg validation and output shaping.
// validate() functions throw ToolValidationError with a model-facing message.

import { ToolValidationError } from "./types.js";

export function reqString(args: Record<string, unknown>, key: string): string {
  const v = args[key];
  if (typeof v !== "string" || v === "") {
    throw new ToolValidationError(`"${key}" is required and must be a non-empty string.`);
  }
  return v;
}

export function optString(args: Record<string, unknown>, key: string): string | undefined {
  const v = args[key];
  if (v === undefined || v === null) return undefined;
  if (typeof v !== "string") throw new ToolValidationError(`"${key}" must be a string.`);
  return v;
}

export function optNumber(args: Record<string, unknown>, key: string): number | undefined {
  const v = args[key];
  if (v === undefined || v === null) return undefined;
  if (typeof v === "number" && Number.isFinite(v)) return v;
  // Be tolerant: models often send numbers as strings inside JSON.
  if (typeof v === "string" && v.trim() !== "" && Number.isFinite(Number(v))) return Number(v);
  throw new ToolValidationError(`"${key}" must be a number.`);
}

export function optBool(args: Record<string, unknown>, key: string): boolean | undefined {
  const v = args[key];
  if (v === undefined || v === null) return undefined;
  if (typeof v === "boolean") return v;
  if (v === "true") return true;
  if (v === "false") return false;
  throw new ToolValidationError(`"${key}" must be a boolean.`);
}

// Writes are restricted to the project's writable subtrees, so the model can't
// clobber config/lockfiles. The jail enforces this too, but checking here gives
// the model a clear message instead of an opaque jail rejection.
const WRITABLE_PREFIXES = ["src/", "public/", "out/"];

// Normalize a relative path and reject escapes (absolute or "..") shared by both
// the read and write checks.
function normalizeRel(p: string): string {
  const norm = p.replace(/\\/g, "/").replace(/^\.\//, "");
  if (norm.startsWith("/") || norm.includes("..")) {
    throw new ToolValidationError(`Path "${p}" must be relative to the project root with no "..".`);
  }
  return norm;
}

// For write/edit: must land in a writable subtree.
export function assertProjectPath(p: string): string {
  const norm = normalizeRel(p);
  if (!WRITABLE_PREFIXES.some((pre) => norm.startsWith(pre))) {
    throw new ToolValidationError(
      `Path "${p}" must be under one of: ${WRITABLE_PREFIXES.join(", ")}.`,
    );
  }
  return norm;
}

// For read/list: any path inside the project is fair game (package.json,
// tsconfig.json, etc.) so the agent can understand the project — it just can't
// escape the root. The jail still bounds it to the sandbox workdir.
export function assertReadablePath(p: string): string {
  return normalizeRel(p);
}

File: impl/go/internal/tool/validate.go
// Shared arg-validation + path helpers (Go parity of tool/validate.ts).
package tool

import (
	"math"
	"strconv"
	"strings"
)

func reqString(args map[string]any, key string) (string, error) {
	v, ok := args[key]
	if !ok {
		return "", validationErr("%q is required and must be a non-empty string.", key)
	}
	s, ok := v.(string)
	if !ok || s == "" {
		return "", validationErr("%q is required and must be a non-empty string.", key)
	}
	return s, nil
}

func optString(args map[string]any, key string) (string, error) {
	v, ok := args[key]
	if !ok || v == nil {
		return "", nil
	}
	s, ok := v.(string)
	if !ok {
		return "", validationErr("%q must be a string.", key)
	}
	return s, nil
}

// optNumber tolerates models sending numbers as JSON strings.
func optNumber(args map[string]any, key string) (float64, bool, error) {
	v, ok := args[key]
	if !ok || v == nil {
		return 0, false, nil
	}
	switch n := v.(type) {
	case float64:
		if math.IsInf(n, 0) || math.IsNaN(n) {
			return 0, false, validationErr("%q must be a finite number.", key)
		}
		return n, true, nil
	case int:
		return float64(n), true, nil
	case string:
		f, err := strconv.ParseFloat(strings.TrimSpace(n), 64)
		if err != nil {
			return 0, false, validationErr("%q must be a number.", key)
		}
		return f, true, nil
	default:
		return 0, false, validationErr("%q must be a number.", key)
	}
}

func optBool(args map[string]any, key string) (bool, error) {
	v, ok := args[key]
	if !ok || v == nil {
		return false, nil
	}
	switch b := v.(type) {
	case bool:
		return b, nil
	case string:
		if b == "true" {
			return true, nil
		}
		if b == "false" {
			return false, nil
		}
	}
	return false, validationErr("%q must be a boolean.", key)
}

// writablePrefixes mirrors the TS jail: tools may only touch these subtrees.
var writablePrefixes = []string{"src/", "public/", "out/"}

func assertProjectPath(p string) (string, error) {
	norm := strings.ReplaceAll(p, "\\", "/")
	norm = strings.TrimPrefix(norm, "./")
	if strings.HasPrefix(norm, "/") || strings.Contains(norm, "..") {
		return "", validationErr("Path %q must be relative to the project root with no \"..\".", p)
	}
	for _, pre := range writablePrefixes {
		if strings.HasPrefix(norm, pre) {
			return norm, nil
		}
	}
	return "", validationErr("Path %q must be under one of: %s.", p, strings.Join(writablePrefixes, ", "))
}

// assertReadablePath is the looser check for read-only tools (read): any path
// relative to the project root is allowed (e.g. package.json, tsconfig.json) so
// the agent can understand the project, but absolute paths and ".." escapes are
// still rejected. Write/edit keep the stricter assertProjectPath.
func assertReadablePath(p string) (string, error) {
	norm := strings.ReplaceAll(p, "\\", "/")
	norm = strings.TrimPrefix(norm, "./")
	if strings.HasPrefix(norm, "/") || strings.Contains(norm, "..") {
		return "", validationErr("Path %q must be relative to the project root with no \"..\".", p)
	}
	return norm, nil
}
